cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+      );
+
+      try {
+        Collection<MetricTimeSeries> metrics = metricStore.query(metricDataQuery);
+        if (metrics == null || metrics.isEmpty()) {
+          // Data is not yet available, which means no data has been ingested by the stream yet
+          return new StreamSize(0L, System.currentTimeMillis());
+        }
+
+        MetricTimeSeries metric = metrics.iterator().next();
+        List<TimeValue> timeValues = metric.getTimeValues();
+        if (timeValues == null || timeValues.size() != 1) {
+          throw new IOException("Should collect exactly one time value");
+        }
+        TimeValue timeValue = timeValues.get(0);
+        // The metric store gives us 0 as the timestamp, hence we cannot use it here
+        return new StreamSize(timeValue.getValue(), System.currentTimeMillis());
+      } catch (Exception e) {
+        Throwables.propagateIfInstanceOf(e, IOException.class);
+        throw new IOException(e);
+      }
+    }
+  }
+
+  /**
+   * Wrapper around a {@link StreamSizeSchedule} which will run a program whenever it receives enough
+   * data from a stream, via notifications.
+   */
+  private final class StreamSizeScheduleTask {
+    private final Id.Program programId;
+    private final SchedulableProgramType programType;
+    private final AtomicBoolean active;
+    private StreamSizeSchedule streamSizeSchedule;
+
+    // Size, in bytes, given by the polling info which serves as a base when comparing polling info
+    private long basePollSize;
+    // Time, in milliseconds, when the previous attribute was computed
+    private long basePollTs;
+
+    // Size, in bytes, of the stream at the last recorded execution
+    private long lastRunSize;
+    // Logical time, in milliseconds, of the last recorded execution
+    private long lastRunTs;
+
+    private StreamSizeScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                   StreamSizeSchedule streamSizeSchedule) {
+      this.programId = programId;
+      this.programType = programType;
+      this.streamSizeSchedule = streamSizeSchedule;
+      this.active = new AtomicBoolean(false);
+    }
+
+    /**
+     * Start a stream size schedule task.
+     *
+     * @param basePollSize base size of the stream to start counting from. This info comes from polling the stream
+     * @param basePollTs time at which the {@code basePollSize} was obtained
+     * @param active {@code true} if the schedule is active, {@code false} if it is suspended
+     * @param persist {@code true} if this information should be persisted, {@code false} otherwise
+     */
+    public void startSchedule(long basePollSize, long basePollTs, boolean active, boolean persist) {
+      LOG.debug("Starting schedule {} with basePollSize {}, basePollTs {}, active {}. Should be persisted: {}",
+                streamSizeSchedule.getName(), basePollSize, basePollTs, active, persist);
+      this.basePollSize = basePollSize;
```

Comment:
So, just to understand this. It seems that basePollSize/Ts is the size and time when the task was last triggered. And lastRunTs/Size is when the task actually was executed. Is that correct? 

If so, then I am not sure whether this initialization is correct. If basePollSize/Ts are -1, then this is a new task that is scheduled for the first time. If not, then they represent what was persisted in the store. So if the task has run before and that was persisted, then why are you setting lastRunTs/Size to -1?